Run cargo fmt in CI

tcharding opened the pull request:
Run the formatter in CI so we stop continually introducing formatting problems in the code that is currently supposed to be formatted.

apoelstra reviewed (approved):
ACK b00e275de0bbe2ebb12ab59e4fb99ac3ce5b1fe4

tcharding commented on contrib/test.sh:
```diff
@@ -80,15 +88,12 @@ then
+            echo "RUSTUP_TOOLCHAIN is set to a stable toolchain but DO_BENCH requires a non-stable (beta, nightly) toolchain"
         else
-            echo "DO_BENCH requires a nightly toolchain"
+            echo "DO_BENCH requires a nightly a non-stable (beta, nightly) toolchain"
```
sloppy work, this needs fixing. Will wait for any other review before doing it.

apoelstra replied:
I don't think it's so bad. Bash is not a super expressive language (or at least, its advanced features are completely opaque :))

tcharding replied:
Oh I just meat the English is bad.

apoelstra reviewed (dismissed):
ACK 7d603fac2df9d8699168db8164e8fcfebc31693a

Kixunil commented on contrib/test.sh:
```diff
@@ -16,11 +10,25 @@ fi
+        exit 1
+    fi
+    rustup component add rustfmt
+    cargo fmt --check
```
Missing `|| exit 1`? I don't see `set -e` anywhere.

Kixunil replied:
Oh, didn't notice that one, from `man sh`:

> -e errexit       If not interactive, exit immediately if any untested command fails.  The exit status of a command is considered to be explicitly tested if the command is used to control an if,
>                            elif, while, or until; or if the command is the left hand operand of an “&&” or “||” operator.

I'm more familiar with `set -e` and suspect it might be clearer so it'd be nice to change it if I'm correct but not very important.

tcharding replied:
Ha!, I stumbled on this also and assumed that some Bash wizard had written it and it was just me that hadn't seen it before :)

Will change.

Kixunil reviewed (commented):
Looks buggy.

Kixunil reviewed (approved):
ACK 7d603fac2df9d8699168db8164e8fcfebc31693a

dunxen reviewed (dismissed):
ACK 7d603fa

apoelstra reviewed (approved):
ACK 83cd37c61d11c6dc91f6ba3deaad71b7cbb10c39

tcharding commented:
I ran `shellcheck` on this branch and found a single bug, fixed and pushed as a separate patch.

apoelstra reviewed (approved):
ACK ac62fa7cc6fd32d4b417fda9159cad06431e735d

tcharding commented:
You know it might actually be better to hold off on merging this until we have stopped ignoring submodules, that way when we are moving things around we can move them in a separate commit to formatting them and still have every commit pass CI.

apoelstra commented:
I think, no matter what we'll have to adjust the ignore rules while moving stuff. Enforcing it in CI, vs on people's local machines (where I believe some contributors have configured their editors to auto-mangle code according to rustfmt) doesn't seem to make a lot of difference.

tcharding commented:
> to auto-mangle code according to rustfmt

I had to laugh at this one :)

tcharding commented:
Changes in force push:
- Rebase (remove changes to the ci script that have been merged in other PRs)

apoelstra reviewed (dismissed):
ACK 570b1c9078ab7d756aa646c4e332fc90ba04d46d

tcharding commented:
Rebased to pick up formatting fixes just merged. No other changes.

apoelstra reviewed (approved):
ACK a9d298a233dbe96ecdc3bf51f26f250c93088dbc

tcharding commented:
Rebase and add initial patch to fix up trivial formatting issue.

dunxen reviewed (approved):
ACK 899e63c

apoelstra reviewed (dismissed):
ACK 899e63c21b9f32b413949f5e481ec4ee0573a623

tcharding commented:
Changes in force push:
- Rebased and added an initial patch that runs the formatter
- Reworded commit `521d4688 Improve the DO_BENCH error message` (just removed `ci: ` prefix from the brief description).

apoelstra reviewed (dismissed):
ACK 55716161abc2334670252f6768a6e226cfe3e0e3

tcharding commented on bitcoin/contrib/test.sh:
```diff
@@ -94,16 +99,24 @@ then
+        exit 1
+    fi
+    rustup component add rustfmt
+    cargo fmt --check
```
This is wrong, should be `cargo fmt --check || exit 1`.

apoelstra replied:
Because we have `set -e` at the top I don't think this is necessary. Though I wouldn't mind it either, since it makes things clearer.

tcharding replied:
oh yes, of course. My bad.

tcharding replied:
I left it as is because non of the other commands explicitly exit, changing this would require me to change all instances across all our CI scripts else I might start twitching when I next read them.

tcharding commented:
Needs rebase :)

tcharding commented:
Changes in force push:
- rebase and redo first patch to include the few trivial formatting issues currently on master.

apoelstra reviewed (dismissed):
ACK 780554f7af1a34b123ee48b1ce2a7d32794333f0

dunxen reviewed (dismissed):
ACK 780554f

tcharding commented:
@apoelstra (or @sanket1729), in case you are wondering, this can be merged any time - don't worry about conflicts with other formatting things, I can rebase.

Changes since @Kixunil last ack'ed are rebases and some changes to the `bitcoin/contrib/test.sh` script. I believe all review suggestions have been addressed.

tcharding commented:
Gentle bump please team, lets get this in to stop the formatting issues from continually creeping into `master`.

tcharding commented:
Force push is just rebase and redo of the initial "run formatter" patch.

apoelstra reviewed (dismissed):
ACK d796dc30eb63c5e4f7a18bbc8a3c57e0a2c63f78

apoelstra commented on bitcoin/contrib/test.sh:
```diff
@@ -12,11 +12,16 @@ fi
+    STABLE=false
     NIGHTLY=true
 fi
+if cargo --version | gre beta; then
```
Actually no, because this is in an `if` command it is probably just always failing, so the `if` is never taken.

But I don't know bash well enough to be certain.

Kixunil replied:
@apoelstra is correct.

apoelstra commented on bitcoin/contrib/test.sh:
```diff
@@ -12,11 +12,16 @@ fi
     NIGHTLY=true
 fi
+if cargo --version | gre beta; then
+    STABLE=false
```
But our MSRV *is* stable, for purposes of this check. "stable" meaning that it won't allow compiling unstable features.

But I see, there are actually two notions of "stable" being used simultaneously by rustup ... one means "the current release" and the other means "a compiler that won't compile unstable stuff" :(

tcharding replied:
`$NIGHTLY` and `$STABLE` are only used to improve the error messages.

tcharding commented:
Changes in force push:
- Rebased
- Added fixes to new formatting issues in patch 1
- Fixed the `gre` -> `grep`

apoelstra commented:
IDK what to make of this asan failure. It doesn't have enough information to zero in on it, and the error message suggests it was a fluke or a bug in our tooling ... but I still don't feel comfortable resetting CI.

sanket1729 commented:
I am not able to reproduce this locally. I did try running `DO_ASAN=true ./contrib/test.sh` inside `rust-bitcoin/hashes` 10 times.

apoelstra commented:
I just kicked it. Let's not knock ourselves out tracking down a probably-spurious asan error in safe code with no backtrace information in a PR that has nothing to do with it.

apoelstra reviewed (approved):
ACK 3919ceda883f20238884929a149b5fc81e6d4cfb

apoelstra commented:
I also can't reproduce it locally, even with the latest nightly (from last night).

tcharding commented:
I cannot reproduce locally either, I checked all tool versions were the same and ran the memory sanitizer command in a loop a bunch of times.

tcharding commented:
BOOM, reproduced it! I ran the following in `bitcoin` _not_ in `hashes`
```
 CC='clang -fsanitize=memory -fno-omit-frame-pointer'                                         \                                                                         ✘ 1 
  RUSTFLAGS='-Zsanitizer=memory -Zsanitizer-memory-track-origins -Cforce-frame-pointers=yes'   \
  cargo +nightly test --lib --no-default-features --all-features -Zbuild-std --target x86_64-unknown-linux-gnu
```

tcharding commented:
Exactly _why_ the RUSTFLAGS from `hashes` build are being used to run tests in `bitcoin` I haven't worked out though ...

tcharding commented:
I've tracked it down to libc
```
$ pwd                                                                                                                                                                       ✘ 101 
/home/tobin/build/github.com/rust-lang/libc
$ cargo clean; CC='clang -fsanitize=memory -fno-omit-frame-pointer' RUSTFLAGS='-Zsanitizer=memory -Zsanitizer-memory-track-origins -Cforce-frame-pointers=yes' cargo +nightly test 
   Compiling libc v0.2.139 (/home/tobin/build/github.com/rust-lang/libc)
error: failed to run custom build command for `libc v0.2.139 (/home/tobin/build/github.com/rust-lang/libc)`

Caused by:
  process didn't exit successfully: `/home/tobin/.cache/cargo/debug/build/libc-a5354087f4621f19/build-script-build` (exit status: 1)
  --- stdout
  cargo:rerun-if-changed=build.rs

  --- stderr
  ==148088==WARNING: MemorySanitizer: use-of-uninitialized-value
      #0 0x5586dd2ffa72  (/home/tobin/.cache/cargo/debug/build/libc-a5354087f4621f19/build-script-build+0x9ea72) (BuildId: bbbfe69bfe3d26afe9b2499f55a0b80631b2071a)
      #1 0x5586dd2fd1cc  (/home/tobin/.cache/cargo/debug/build/libc-a5354087f4621f19/build-script-build+0x9c1cc) (BuildId: bbbfe69bfe3d26afe9b2499f55a0b80631b2071a)
      #2 0x5586dd30e1f3  (/home/tobin/.cache/cargo/debug/build/libc-a5354087f4621f19/build-script-build+0xad1f3) (BuildId: bbbfe69bfe3d26afe9b2499f55a0b80631b2071a)
      #3 0x5586dd30668c  (/home/tobin/.cache/cargo/debug/build/libc-a5354087f4621f19/build-script-build+0xa568c) (BuildId: bbbfe69bfe3d26afe9b2499f55a0b80631b2071a)
      #4 0x5586dd30d4f8  (/home/tobin/.cache/cargo/debug/build/libc-a5354087f4621f19/build-script-build+0xac4f8) (BuildId: bbbfe69bfe3d26afe9b2499f55a0b80631b2071a)
      #5 0x5586dd33cb6b  (/home/tobin/.cache/cargo/debug/build/libc-a5354087f4621f19/build-script-build+0xdbb6b) (BuildId: bbbfe69bfe3d26afe9b2499f55a0b80631b2071a)
      #6 0x5586dd30d309  (/home/tobin/.cache/cargo/debug/build/libc-a5354087f4621f19/build-script-build+0xac309) (BuildId: bbbfe69bfe3d26afe9b2499f55a0b80631b2071a)
      #7 0x5586dd3031be  (/home/tobin/.cache/cargo/debug/build/libc-a5354087f4621f19/build-script-build+0xa21be) (BuildId: bbbfe69bfe3d26afe9b2499f55a0b80631b2071a)
      #8 0x7f0534ab4d8f  (/lib/x86_64-linux-gnu/libc.so.6+0x29d8f) (BuildId: 69389d485a9793dbe873f0ea2c93e02efaa9aa3d)
      #9 0x7f0534ab4e3f  (/lib/x86_64-linux-gnu/libc.so.6+0x29e3f) (BuildId: 69389d485a9793dbe873f0ea2c93e02efaa9aa3d)
      #10 0x5586dd271874  (/home/tobin/.cache/cargo/debug/build/libc-a5354087f4621f19/build-script-build+0x10874) (BuildId: bbbfe69bfe3d26afe9b2499f55a0b80631b2071a)

    Uninitialized value was created by an allocation of '_2' in the stack frame of function '_ZN18build_script_build19rustc_minor_nightly17ha05947cc82f4f00dE'
      #0 0x5586dd2ff680  (/home/tobin/.cache/cargo/debug/build/libc-a5354087f4621f19/build-script-build+0x9e680) (BuildId: bbbfe69bfe3d26afe9b2499f55a0b80631b2071a)

  SUMMARY: MemorySanitizer: use-of-uninitialized-value (/home/tobin/.cache/cargo/debug/build/libc-a5354087f4621f19/build-script-build+0x9ea72) (BuildId: bbbfe69bfe3d26afe9b2499f55a0b80631b2071a) 
  Exiting
```

tcharding commented:
Raised an issue on libc: https://github.com/rust-lang/libc/issues/3075

tcharding commented:
Sanitizer fixed in #1566

sanket1729 commented:
Is this 0.30 blocking? We can easily merge formating changes post release.

tcharding commented:
Not blocking 0.30.0 - I was hoping to slip this through at some stage, ether after release like you say or right at the end of the feature freeze before release if the remaining PRs to merge are only mine (so I'm the only one to have to go through the rebase pain :)

tcharding commented:
Rebase only, no other changes.

tcharding commented:
Rebased on top of last nights merges. Added new formatting changes to commit: `c1220e26 Run the formatter`

tcharding commented:
Rebased so I can use this as the base of #1434 still. No changes needed.

tcharding commented:
Lets push this one in crew, there have been plenty of acks in the past so the concept is good. Its only commit `6ee234cb Run the formatter` that has change with all the force pushes.

Kixunil commented:
IMO we should at least get #1533 in before this. I'd hate to re-review it after formatting changes.

tcharding commented:
Sure thing, my bad I didn't think of that one.

tcharding commented:
Draft till #1533 goes in

apoelstra commented:
Needs rebase, then let's gooooo.

tcharding commented:
Rebased and re-did `a11cf075 Run the formatter`

apoelstra commented:
This only enables the formatter in CI for bitcoin, not the other crates. Is that intentional?

apoelstra reviewed (dismissed):
ACK 65eee20c53593763b2b8953a49a7ae752c571322

tcharding commented:
> This only enables the formatter in CI for bitcoin, not the other crates. Is that intentional?

Not intentional, my mistake. Fixed and force pushed. FTR we still ignore `hashes` in `rustfmt.toml`  - that is intentionally not fixed here. 

Thanks

apoelstra commented:
Looks good -- though we're still not covering `internals`. I'd suggest adding the formatting code to the top-level `contrib/test.sh` rather than to the individual crates' `contrib/test.sh`s.

sanket1729 commented:
Is this the only 0.30.0 blockcing PR?

tcharding commented:
Please see https://github.com/rust-bitcoin/rust-bitcoin/pull/1699 for release blocking PRs

tcharding commented:
> Looks good -- though we're still not covering `internals`. 

Sure we are, its in `c1360067 Enable formatting in CI`

I'd suggest adding the formatting code to the top-level `contrib/test.sh` rather than to the individual crates' `contrib/test.sh`s.

I'm not comfortable doing that for this PR because currently the `rust-bitcoin/contrib/test.sh` is just a thin wrapper. If we want to put othe logic into it we should put _all_ the logic that makes sense into it.

apoelstra reviewed (approved):
ACK 2d6467f980276f83d244551ba7432ac53610b7a8

sanket1729 reviewed (approved):
ACK 2d6467f980276f83d244551ba7432ac53610b7a8

apoelstra commented:
We have an old ACK from Kix and we've merged everything we wanted to merge before this. I'm gonna do it.
